File: 1-50/3/answer.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Answer 
{
    private static int getNextPrime(int prime, List<int> primes) 
    {
        var number = prime + 1;
        while(primes.Any(p => number % p == 0))
        {
            ++number;
        }
        return number;
    }

    static void Main()
    {
        long target = 600851475143L;
        var maxPrime = target;
        var prime = 2;
        var primes = new List<int>(100);
        primes.Add(prime);

        while(prime * prime <= target)
        {
            while(target % prime == 0)
            {
                target /= prime;
                maxPrime = prime;
            }

            prime = getNextPrime(prime, primes);
            primes.Add(prime);
        }

        Console.WriteLine(maxPrime);
    }

}

File: 1-50/3/answer.kt
import kotlin.collections.List
import kotlin.collections.MutableList


fun getNextPrime(primes: List<Long>): Long {
    var number = primes.last() + 1L
    while(primes.any{ number % it == 0L }) {
        number += 1L
    }
    return number
}

fun main() {
    var target = 600851475143L
    var maxPrime = target
    var primes = mutableListOf<Long>(2L)

    while(primes.last() * primes.last() <= target) {
        while(target % primes.last() == 0L) {
            target /= primes.last()
            maxPrime = primes.last()
        }

        primes.add(getNextPrime(primes))
    }

    println(maxPrime)
}